compiler/rustc_codegen_llvm/src/intrinsic.rs

```diff
@@ -203,6 +203,42 @@ impl<'ll, 'tcx> IntrinsicCallMethods<'tcx> for Builder<'_, 'll, 'tcx> {
             }
             sym::unlikely => self
                 .call_intrinsic("llvm.expect.i1", &[args[0].immediate(), self.const_bool(false)]),
+            sym::select_unpredictable => {
+                let cond = args[0].immediate();
+                assert_eq!(args[1].layout, args[2].layout);
+                let select = |bx: &mut Self, true_val, false_val| {
+                    let result = bx.select(cond, true_val, false_val);
+                    bx.set_unpredictable(&result);
+                    result
+                };
+                match (args[1].val, args[2].val) {
+                    (OperandValue::Ref(true_val), OperandValue::Ref(false_val)) => {
+                        assert!(true_val.llextra.is_none());
+                        assert!(false_val.llextra.is_none());
+                        let llty = if let PassMode::Cast { cast: ty, pad_i32: _ } = &fn_abi.ret.mode
+                        {
+                            ty.llvm_type(self)
+                        } else {
+                            args[1].layout.llvm_type(self)
+                        };
+                        let true_val = self.load(llty, true_val.llval, true_val.align);
+                        let false_val = self.load(llty, false_val.llval, false_val.align);
+                        let out = select(self, true_val, false_val);
+                        if !result.layout.is_zst() {
+                            self.store_to_place(out, result.val);
+                        }
+                        return Ok(());
+                    }
```

Comment:
I'm not 100% sure on the logic for `OperandValue::Ref`, I mostly copied it off `volatile_load` below.

Reply:
Copying anything from the volatile_load implementation is always wrong :)

I think for Ref it would make the most sense to not load the values in the first place and perform the select on the pointers instead.

Reply:
I tried that at first but then LLVM wasn't preserving the `!unpredictable` on the select.

Reply:
That's something we should fix in LLVM. Do you have the pre-optimization IR you generated?

Reply:
Actually I misread the IR. It's fine either way. I'll change it to select the pointer.